File: BudgetTracker.cs
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;

namespace BudgetTracker
{
    class BudgetTracker
    {
        static string dataFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "data.json");
        static List<Transaction> transactions = new List<Transaction>();
        static decimal savingsGoal = 0;

        static void Main(string[] args)
        {
            LoadData();

            while (true)
            {
                Console.WriteLine("\nWelcome to the Personal Budget Tracker!");
                Console.WriteLine("Please select an option:");
                Console.WriteLine("1. Add Income");
                Console.WriteLine("2. Add Expense");
                Console.WriteLine("3. View Monthly Summary");
                Console.WriteLine("4. Set Savings Goal");
                Console.WriteLine("5. Exit\n");

                Console.Write("Enter your choice: ");
                var choice = Console.ReadLine();

                switch (choice)
                {
                    case "1":
                        AddTransaction("income");
                        break;
                    case "2":
                        AddTransaction("expense");
                        break;
                    case "3":
                        ViewMonthlySummary();
                        break;
                    case "4":
                        SetSavingsGoal();
                        break;
                    case "5":
                        SaveData();
                        Console.WriteLine("Goodbye!");
                        return;
                    default:
                        Console.WriteLine("Invalid option. Please try again.");
                        break;
                }
            }
        }

        static void AddTransaction(string type)
        {
            Console.Write("Enter category: ");
            string category = Console.ReadLine() ?? "Uncategorized";

            Console.Write("Enter amount: ");
            decimal amount;
            while (!decimal.TryParse(Console.ReadLine(), out amount) || amount <= 0)
            {
                Console.Write("Invalid amount. Please enter a positive number: ");
            }

            Console.Write("Enter description: ");
            string description = Console.ReadLine() ?? "No description";

            Transaction transaction = new Transaction
            {
                Date = DateTime.Now.ToString("yyyy-MM-dd"),
                Type = type,
                Category = category,
                Amount = amount,
                Description = description
            };

            transactions.Add(transaction);
            Console.WriteLine($"{type} added successfully!");
            SaveData();
        }

        static void ViewMonthlySummary()
        {
            decimal totalIncome = 0;
            decimal totalExpenses = 0;

            foreach (var transaction in transactions)
            {
                if (DateTime.Parse(transaction.Date).Month == DateTime.Now.Month)
                {
                    if (transaction.Type == "income")
                        totalIncome += transaction.Amount;
                    else if (transaction.Type == "expense")
                        totalExpenses += transaction.Amount;
                }
            }

            decimal savings = totalIncome - totalExpenses;
            Console.WriteLine("\nMonthly Summary:");
            Console.WriteLine($"Total Income: {totalIncome:C}");
            Console.WriteLine($"Total Expenses: {totalExpenses:C}");
            Console.WriteLine($"Savings: {savings:C}");

            if (savingsGoal > 0)
            {
                Console.WriteLine($"Savings Goal: {savingsGoal:C}");
                Console.WriteLine($"Progress: {(savings / savingsGoal * 100):F2}%");
            }

            if (totalExpenses > totalIncome)
                Console.WriteLine("Warning: Your expenses exceed your income!");
        }

        static void SetSavingsGoal()
        {
            Console.Write("Enter your savings goal: ");
            while (!decimal.TryParse(Console.ReadLine(), out savingsGoal) || savingsGoal <= 0)
            {
                Console.Write("Invalid goal. Please enter a positive number: ");
            }

            Console.WriteLine($"Savings goal set to {savingsGoal:C}");
        }

        static void LoadData()
        {
            if (File.Exists(dataFile))
            {
                try
                {
                    string json = File.ReadAllText(dataFile);
                    transactions = JsonConvert.DeserializeObject<List<Transaction>>(json) ?? new List<Transaction>();
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error loading data: {ex.Message}");
                }
            }
        }

        static void SaveData()
        {
            try
            {
                string json = JsonConvert.SerializeObject(transactions, Formatting.Indented);
                File.WriteAllText(dataFile, json);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error saving data: {ex.Message}");
            }
        }

        class Transaction
        {
            public string Date { get; set; } = string.Empty;
            public string Type { get; set; } = string.Empty;
            public string Category { get; set; } = string.Empty;
            public decimal Amount { get; set; }
            public string Description { get; set; } = string.Empty;
        }
    }
}

File: BudgetTracker.py
import os
import json
from datetime import datetime

class BudgetTracker:
    data_file = os.path.join(os.path.dirname(_file_), 'data.json')
    transactions = []
    savings_goal = 0

    @staticmethod
    def main():
        BudgetTracker.load_data()

        while True:
            print("\nWelcome to the Personal Budget Tracker!")
            print("Please select an option:")
            print("1. Add Income")
            print("2. Add Expense")
            print("3. View Monthly Summary")
            print("4. Set Savings Goal")
            print("5. Exit\n")

            choice = input("Enter your choice: ")

            if choice == "1":
                BudgetTracker.add_transaction("income")
            elif choice == "2":
                BudgetTracker.add_transaction("expense")
            elif choice == "3":
                BudgetTracker.view_monthly_summary()
            elif choice == "4":
                BudgetTracker.set_savings_goal()
            elif choice == "5":
                BudgetTracker.save_data()
                print("Goodbye!")
                break
            else:
                print("Invalid option. Please try again.")

    @staticmethod
    def add_transaction(transaction_type):
        category = input("Enter category: ") or "Uncategorized"

        while True:
            try:
                amount = float(input("Enter amount: "))
                if amount <= 0:
                    raise ValueError("Amount must be positive.")
                break
            except ValueError as e:
                print(e)

        description = input("Enter description: ") or "No description"

        transaction = {
            "date": datetime.now().strftime("%Y-%m-%d"),
            "type": transaction_type,
            "category": category,
            "amount": amount,
            "description": description
        }

        BudgetTracker.transactions.append(transaction)
        print(f"{transaction_type.capitalize()} added successfully!")
        BudgetTracker.save_data()

    @staticmethod
    def view_monthly_summary():
        total_income = 0
        total_expenses = 0
        current_month = datetime.now().month

        for transaction in BudgetTracker.transactions:
            transaction_month = datetime.strptime(transaction["date"], "%Y-%m-%d").month
            if transaction_month == current_month:
                if transaction["type"] == "income":
                    total_income += transaction["amount"]
                elif transaction["type"] == "expense":
                    total_expenses += transaction["amount"]

        savings = total_income - total_expenses

        print("\nMonthly Summary:")
        print(f"Total Income: ${total_income:.2f}")
        print(f"Total Expenses: ${total_expenses:.2f}")
        print(f"Savings: ${savings:.2f}")

        if BudgetTracker.savings_goal > 0:
            progress = (savings / BudgetTracker.savings_goal) * 100
            print(f"Savings Goal: ${BudgetTracker.savings_goal:.2f}")
            print(f"Progress: {progress:.2f}%")

        if total_expenses > total_income:
            print("Warning: Your expenses exceed your income!")

    @staticmethod
    def set_savings_goal():
        while True:
            try:
                BudgetTracker.savings_goal = float(input("Enter your savings goal: "))
                if BudgetTracker.savings_goal <= 0:
                    raise ValueError("Savings goal must be positive.")
                break
            except ValueError as e:
                print(e)

        print(f"Savings goal set to ${BudgetTracker.savings_goal:.2f}")

    @staticmethod
    def load_data():
        if os.path.exists(BudgetTracker.data_file):
            try:
                with open(BudgetTracker.data_file, 'r') as file:
                    BudgetTracker.transactions = json.load(file)
            except Exception as e:
                print(f"Error loading data: {e}")

    @staticmethod
    def save_data():
        try:
            with open(BudgetTracker.data_file, 'w') as file:
                json.dump(BudgetTracker.transactions, file, indent=4)
        except Exception as e:
            print(f"Error saving data: {e}")

if _name_ == "_main_":
    BudgetTracker.main()
